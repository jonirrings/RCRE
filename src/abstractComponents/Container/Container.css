.rcre-abstract-container {
    overflow: auto;
    padding: 0 10px;
}