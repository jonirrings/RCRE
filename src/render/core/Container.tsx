import * as React from 'react';
import componentLoader from '../util/componentLoader';
import createElement from '../util/createElement';

export class ContainerProps {
    type: string;
    [s: string]: any
}

export function CreateContainer(info: ContainerProps) {
    let componentInfo = componentLoader.getComponent(info.type);
    
    if (!componentInfo) {
        return null;
    }
    
    let {
        component,
        componentInterface
    } = componentInfo;
    
    const wrappedComponentName = component.displayName || component.name || 'ContainerComponent';
    const displayName = `Container(${wrappedComponentName})`;

    class Container extends React.Component<ContainerProps, {}> {
        static WrappedComponent: string;
        static displayName: string;

        constructor() {
            super();
        }

        render() {
            if (!component) {
                return null;
            }
            
            return createElement<ContainerProps>(component, componentInterface, info);
        }
    }

    Container.WrappedComponent = wrappedComponentName;
    Container.displayName = displayName;
    
    return React.createElement<ContainerProps>(Container, Object.assign(info, {
        key: info.key
    }));
}