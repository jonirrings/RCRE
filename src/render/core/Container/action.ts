export const SET_DATA = 'SET_DATA';
export const INIT_DATA = 'INIT_DATA';
export type SET_DATA_PAYLOAD = {
    type: string;
    newValue: any;
};
export type INIT_DATA_PAYLOAD = any;

export type IActions = {
    SET_DATA: { 
        type: typeof SET_DATA,
        payload: SET_DATA_PAYLOAD 
    },
    INIT_DATA: {
        type: typeof INIT_DATA,
        payload: INIT_DATA_PAYLOAD
    }
};

export type IAction = IActions[keyof IActions];

export const actionCreators = {
    setData: (payload: SET_DATA_PAYLOAD) => ({
        type: SET_DATA as typeof SET_DATA,
        payload
    }),
    initData: (payload: INIT_DATA_PAYLOAD) => ({
        type: INIT_DATA as typeof INIT_DATA,
        payload
    })
};