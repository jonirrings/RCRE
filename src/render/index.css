.rcre-render {
    width: 100%;
    background: transparent;
}

.rcre-spin {
    width: 100%;
}