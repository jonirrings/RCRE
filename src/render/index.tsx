import * as React from 'react';
import Page, {PageProps} from './core/Page';

interface RenderPropsInterface {
    code: string;
}

class Render extends React.Component<RenderPropsInterface, {}> {
    constructor() {
        super();
    }

    shouldComponentUpdate(nextProps: RenderPropsInterface, nextState: {}) {
        try {
            JSON.parse(nextProps.code);
            return true;
        } catch (e) {
            // TODO Error Report
            return false;
        }
    }

    render() {
        let info: PageProps = JSON.parse(this.props.code);

        return (
            <div className="render">
                <Page
                    title={info.title}
                    body={info.body}
                />
            </div>
        );
    }
}

export default Render;