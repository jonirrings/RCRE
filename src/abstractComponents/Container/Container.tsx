import * as React from 'react';
import * as _ from 'lodash';
import {BasicConfig, BasicContainer, BasicContainerPropsInterface} from '../../render/core/Container/types';
import * as PropTypes from 'prop-types';
import {Validate} from 'class-validator';
import {IsPageInfo} from '../../render/util/validators';
import {isExpression, runInContext} from '../../render/util/vm';
import {createChild} from '../../render/util/createChild';

import './Container.css';

export class ContainerConfig extends BasicConfig {
    style?: React.CSSProperties;
    
    children?: BasicConfig[];
}

export class ContainerPropsInterface extends BasicContainerPropsInterface {
    @Validate(IsPageInfo, [ContainerConfig])
    info: ContainerConfig;
}

export default class AbstractContainer extends BasicContainer<ContainerPropsInterface, {}> {
    static childContextTypes = {
        abstractContainer: PropTypes.bool
    };
    
    constructor() {
        super();

        this.handleChange = this.handleChange.bind(this);
    }

    getChildContext() {
        return {
            // 抽象Abstract组件, 为子级的所有组件提供一致的数据模型
            abstractContainer: true
        };
    }

    render() {
        let children;
        
        if (Array.isArray(this.props.info.children)) {
            let ret = this.parseChildrenExpression(this.props.info.children);
            children = ret.map((child, index) => {
                if (child.hidden) {
                    return React.createElement('div', {
                        style: {
                            display: 'none'
                        },
                        key: `${child.type}_${0}_${index}`
                    });
                }
                
                return this.renderChild(child, 0, index);
            });
        }
        
        let childElement = React.createElement('div', {
            style: this.props.info.style,
            className: 'rcre-abstract-container'
        }, children);
        
        return this.renderChildren(childElement);
    }

    private renderChild(info: BasicConfig, depth: number, index: number) {
        console.log(info);
        return createChild(info, {
            key: `${info.type}_${depth}_${index}`,
            info: info,
            onChange: this.handleChange,
            $data: this.props.$data,
            $setData: this.props.$setData,
            $setDataList: this.props.$setDataList
        }, this.context.form, false);
    }

    handleChange(type: string, val: any) {
        this.props.onChange(type, val);
    }

    private parseChildrenExpression(info: BasicConfig[]) {
        let infoCopy = _.cloneDeep(info);
        let mirror = this.props.$data.toObject();
        let self = this;

        if (_.isEmpty(mirror)) {
            return infoCopy;
        }
        
        const validParseConfig = [
            'children',
            'data'
        ];

        function parseExpression(reference: Object, val: any, name: string | number) {
            if (isExpression(val)) {
                let ret = runInContext(val, {
                    $data: mirror,
                    $global: self.context.$global
                });
                
                if (!_.isNil(ret)) {
                    reference[name] = ret;
                }
            }

            if (val && typeof name === 'string' && validParseConfig.indexOf(name) >= 0) {
                reference[name] = self.parseChildrenExpression(val);
            }
        }
        
        _.each(infoCopy, (config, index) => {
            if (isExpression(config)) {
                parseExpression(infoCopy, config, index);
                return;
            }
            
            if (_.isPlainObject(config)) {
                _.each(config, (val, name) => {
                    parseExpression(config, val, name);
                });
            }
        });

        return infoCopy;
    }
}