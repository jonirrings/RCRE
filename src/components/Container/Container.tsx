import * as React from 'react';
import {BasicConfig, BasicContainer, BasicContainerPropsInterface} from '../../render/core/Container/types';
import {Validate} from 'class-validator';
import {IsPageInfo} from '../../render/util/validators';
import './Container.css';
import {createChild} from '../../render/util/createChild';
import Container from '../../render/core/Container/index';
import {ProviderSourceConfig} from '../../render/core/DataProvider/Controller';
import componentLoader from '../../render/util/componentLoader';
import {CustomerSourceConfig} from '../../render/core/DataCustomer/Controller';

export class ContainerConfig extends BasicConfig {
    /**
     * 字级组件
     */
    children: BasicConfig[];

    /**
     * 数据模型Key
     */
    model: string;

    /**
     * dataProvider配置
     */
    dataProvider?: ProviderSourceConfig | ProviderSourceConfig[];

    /**
     * dataCustomer配置
     */
    dataCustomer?: CustomerSourceConfig;
}

export class ContainerPropsInterface extends BasicContainerPropsInterface {
    @Validate(IsPageInfo, [ContainerConfig])
    info: ContainerConfig;
}

export default class AbstractContainer extends BasicContainer<ContainerPropsInterface, {}> {
    constructor() {
        super();
    }

    render() {
        let children;
        
        if (Array.isArray(this.props.info.children)) {
            children = this.props.info.children.map((child, index) => {
                return this.renderChild(child, 0, index);
            });
        }
        
        let childElement = React.createElement(Container, this.props, children);
        
        return this.renderChildren(this.props.info, childElement);
    }

    private renderChild(info: BasicConfig, depth: number, index: number) {
        return createChild(info, {
            key: `${info.type}_${depth}_${index}`,
            info: info
        });
    }
}

componentLoader.addComponent('container', AbstractContainer, ContainerPropsInterface);