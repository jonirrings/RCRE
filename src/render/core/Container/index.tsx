import * as React from 'react';
import * as _ from 'lodash';
import {BasicContainer, BasicContainerPropsInterface, ContainerProps} from './types';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import {actionCreators, IAction, SET_DATA_LIST_PAYLOAD} from './action';
import {RootState} from '../../data/reducers';
import {Map} from 'immutable';
// import ParamsInjector from '../../util/injector';
// // import {Spin} from 'antd';
import {compileValueExpress, isExpression, keepExpressionData} from '../../util/vm';
// import {request} from '../../services/api';
// import {AxiosRequestConfig, AxiosResponse} from 'axios';
// import {compileTimeExpression} from '../../util/dateTime';
import {createChild} from '../../util/createChild';

// First Init Life Circle:
// ComponentWillMount -> Render -> ComponentDidMount

// Component Update Life Circle: 
// componentWillReceiveProps -> shouldComponentUpdate -> ComponentWillUpdate -> Render -> ComponentDidMount
class Container extends BasicContainer<ContainerProps, {}> {
    static WrappedComponent: string;
    static displayName: string;

    // private parseProperty = {};
    // private prevRequestData = {};

    constructor() {
        super();
        // this.loadData = this.loadData.bind(this);
        this.handleChange = this.handleChange.bind(this);
    }

    componentWillMount() {
        if (this.props.info.model) {
            if (!this.props.info.data) {
                this.props.info.data = {};
            }

            // this.props.info.data is readonly

            this.syncInitData(this.props.info.model, this.props);

            // this.setDataIntoStore(this.props.info.model, this.props);
            // if (this.props.info.initialLoad && !this.props.info.hidden) {
            //     this.mergeOriginData(this.props);
            // }
        } else {
            console.error('Container Component Should have model property');
        }
    }

    componentWillReceiveProps(nextProps: ContainerProps) {
        if (this.props.$parent !== nextProps.$parent) {
            this.syncUpdateData(this.props.info.model!, nextProps);   
        }
    }

    componentWillUnmount() {
        if (this.props.info.model) {
            this.props.removeData({
                model: this.props.info.model
            });
        }
    }

    // componentWillReceiveProps(nextProps: ContainerProps) {
    //     if (!!this.props.$parent && this.props.info.model && nextProps.$parent !== this.props.$parent) {
    //         this.setDataIntoStore(this.props.info.model, nextProps, true);
    //         setTimeout(() => {
    //             if (this.props.info.initialLoad && !this.props.info.hidden) {
    //                 this.mergeOriginData(this.props);
    //             }
    //         });
    //     }
    // }

    shouldComponentUpdate(nextProps: ContainerProps) {
        return this.props.$data !== nextProps.$data;
    }

    componentDidUpdate() {
        // if (this.props.info.initialLoad) {
        //     this.mergeOriginData(this.props);
        // }
    }

    render() {
        let info = _.cloneDeep(this.props.info);

        if (!info.model) {
            return <div className="err-text">model should defined in container like components</div>;
        }

        if (!info.children) {
            return <div className="err-text">children props must be specific in Container Component</div>;
        }

        // console.log(this.props.info.model, this.props.$data);

        let childElements = info.children.map((child, index) => {
            // let compiled = compileValueExpress<BasicConfig, Object>(child, {
            //     $data: this.props.$data.toObject(),
            //     $global: this.context.$global
            // }, ['data', 'children']);

            return createChild<BasicContainerPropsInterface>(child, {
                key: `${child.type}_${index}`,
                info: child,
                $data: this.props.$data,
                onChange: this.handleChange
            });
        });

        return (
            <div className="rcre-container">
                {childElements}
            </div>
        );
    }

    /**
     * Init data when ComponentWillMount
     * This function will call this.props.initData function to init data
     * All expression string will be ignored
     *
     * @param {string} model The Component Model Key
     * @param {ContainerProps} props
     */
    private syncInitData(model: string, props: ContainerProps) {
        if (!props.info.data) {
            props.info.data = {};
        }

        let initData = {};

        _.each(props.info.data, (item, key) => {
            if (!isExpression(item)) {
                initData[key] = item;
            }
        });

        this.props.initData({
            model: model,
            data: initData
        });
    }

    /**
     * Update data When ComponentWillReceiveProps
     * This function will parse the expression string in data property from JSON
     * including $parent and $response
     * Then call this.props.setDataList to update data
     *
     * @param {string} model
     * @param {ContainerProps} props
     */
    private syncUpdateData(model: string, props: ContainerProps) {
        if (!props.info.data) {
            props.info.data = {};
        }

        let updateExpression = keepExpressionData(props.info.data);

        if (!_.isEmpty(updateExpression)) {
            let updateData = compileValueExpress(updateExpression, {
                $parent: props.$parent.toObject()
            });

            let payloads: SET_DATA_LIST_PAYLOAD = [];
            _.each(updateData, (val, name) => {
                payloads.push({
                    type: name,
                    newValue: val
                });
            });
            this.props.setDataList(payloads, model);
        }
    }

    // private setDataIntoStore(model: string, nextProps: ContainerProps, merge?: boolean) {
    //     let initData = {};
    //
    //     let infoData = compileValueExpress(nextProps.info.data, {
    //         $parent: nextProps.$parent.toObject(),
    //         $global: this.context.$global
    //     });
    //
    //     infoData = compileTimeExpression(infoData, this.props.info.$nowFormat);
    //
    //     _.each(infoData, (item, key) => {
    //         if (!isExpression(item)) {
    //             initData[key] = item;
    //         }
    //     });
    //
    //     if (!merge) {
    //         this.props.initData({
    //             model: model,
    //             data: initData
    //         });
    //     } else {
    //         let payloads: SET_DATA_LIST_PAYLOAD = [];
    //         _.each(initData, (val, name) => {
    //             payloads.push({
    //                 type: name,
    //                 newValue: val
    //             });
    //         });
    //         this.props.setDataList(payloads, model);
    //     }
    //
    // }

    // public mergeOriginData(props: ContainerProps) {
    //     let injector = new ParamsInjector(props, this.loadData);
    //
    //     injector.finished((payloads: SET_DATA_PAYLOAD[]) => {
    //         this.props.setData({
    //             type: '$loading',
    //             newValue: false
    //         }, this.props.info.model!);
    //
    //         if (payloads.length > 0) {
    //             this.props.setDataList(payloads, this.props.info.model!);
    //         }
    //     });
    // }
    //
    // private loadData() {
    //     let initialLoad = this.props.info.initialLoad;
    //     let requestConfig: AxiosRequestConfig = {};
    //
    //     if (typeof initialLoad === 'string') {
    //         requestConfig = {
    //             url: initialLoad
    //         };
    //     } else if (typeof initialLoad === 'object') {
    //         let data = initialLoad.data;
    //
    //         _.each(data, (val, name) => {
    //             if (isExpression(val)) {
    //                 this.parseProperty[name] = val;
    //             }
    //         });
    //
    //         if (data) {
    //             let property = _.cloneDeep(this.parseProperty);
    //             let compiledRet = compileValueExpress(property, {
    //                 $data: this.props.$data.toObject(),
    //                 $query: this.context.$query,
    //                 $global: this.context.$global
    //             });
    //
    //             Object.assign(initialLoad.data, compiledRet);
    //         }
    //
    //         filterExpressionData(initialLoad.data);
    //
    //         requestConfig = initialLoad;
    //     }
    //
    //     if (!requestConfig || _.isEqual(requestConfig, this.prevRequestData)) {
    //         return Promise.resolve({});
    //     }
    //
    //     this.prevRequestData = _.cloneDeep(requestConfig);
    //
    //     this.props.setData({
    //         type: '$loading',
    //         newValue: true
    //     }, this.props.info.model!);
    //
    //     return request(requestConfig.url!, requestConfig, this.context.$global.proxyServer)
    //         .then((ret: AxiosResponse) => {
    //             return Promise.resolve(ret);
    //         });
    // }
    
    private handleChange(key: string, value: any) {
        this.props.setData({
            type: key,
            newValue: value
        }, this.props.info.model!);
    }

}

const mapStateToProps = (state: RootState, ownProps: any) => {
    return {
        $data: state.container.get(ownProps.info.model) || Map({})
    };
};

const mapDispatchToProps = (dispatch: Dispatch<IAction>) => bindActionCreators({
    setData: actionCreators.setData,
    setDataList: actionCreators.setDataList,
    initData: actionCreators.initData,
    removeData: actionCreators.removeData
}, dispatch);

const mergeProps = (stateProps: any, dispatchProps: any, ownProps: any): any => {
    let parentProps = ownProps.$data || Map({});

    return Object.assign({}, ownProps, stateProps, dispatchProps, {
        $parent: parentProps
    });
};

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Container);