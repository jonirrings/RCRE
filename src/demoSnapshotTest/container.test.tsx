import * as React from 'react';
import {Render, store} from '../render/index';
import * as renderer from 'react-test-renderer';
// import {Map} from 'immutable';
import {actionCreators} from '../render/core/Container/action';
import {configure, render} from 'enzyme';

let Adapter = require('enzyme-adapter-react-15');

configure({
    adapter: new Adapter()
});

describe('Container SnapShot Test', () => {
    beforeEach(() => {
        store.dispatch(actionCreators.clearData());
    });

    // it('simple snapshot', () => {
    //     const json = {
    //         body: [
    //             {
    //                 type: 'container',
    //                 model: 'textField',
    //                 children: [
    //                     {
    //                         type: 'text',
    //                         text: 'helloworld'
    //                     }
    //                 ]
    //             }
    //         ]
    //     };
    //    
    //     class RCRE extends React.Component {
    //         render() {
    //             return <Render code={JSON.stringify(json)} />;
    //         }
    //     }
    //
    //     const element = <RCRE />;
    //     const component = renderer.create(element);
    //     let tree = component.toJSON();
    //     expect(tree).toMatchSnapshot();
    //
    //     let state: any = store.getState();
    //     expect(state.container).toBeInstanceOf(Map);
    //     expect(state.container.get('textField')).toBeInstanceOf(Map);
    //     expect(state.container.get('textField').size).toBe(0);
    // });
    //
    // it('$parent props parse', () => {
    //     const json = {
    //         body: [
    //             {
    //                 type: 'container',
    //                 model: 'textField',
    //                 data: {
    //                     name: 1
    //                 },
    //                 children: [
    //                     {
    //                         type: 'container',
    //                         model: 'innerTextField',
    //                         data: {
    //                             name: '$parent.name'
    //                         },
    //                         children: [
    //                             {
    //                                 type: 'text',
    //                                 text: 'helloworld'
    //                             }
    //                         ]
    //                     }
    //                 ]
    //             }
    //         ]
    //     };
    //
    //     const element = <Render code={JSON.stringify(json)} />;
    //     const component = renderer.create(element);
    //     let tree = component.toJSON();
    //     expect(tree).toMatchSnapshot();
    //
    //     let state: any = store.getState();
    //     expect(state.container).toBeInstanceOf(Map);
    //     expect(state.container.get('textField')).toBeInstanceOf(Map);
    //     expect(state.container.get('textField').size).toBe(1);
    //     expect(state.container.get('textField').get('name')).toBe(1);
    //     expect(state.container.get('innerTextField').get('name')).toBe(1);
    // });

    it('container child props parse', async () => {
        store.dispatch(actionCreators.clearData());
        const json = {
            body: [
                {
                    type: 'container',
                    model: 'textField',
                    data: {
                        name: 1
                    },
                    children: [
                        {
                            type: 'container',
                            model: 'innerTextField',
                            data: {
                                name: '$parent.name'
                            },
                            children: [
                                {
                                    type: 'text',
                                    text: '$data.text'
                                }
                            ]
                        }
                    ]
                }
            ]
        };

        const element = <Render code={JSON.stringify(json)} />;
        const component = renderer.create(element);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        let state = store.getState();
        expect(state.container.get('textField').get('name')).toBe(1);
        expect(state.container.get('innerTextField').get('name')).toBe(1);

        store.dispatch(actionCreators.clearData());

        const wrapper = render(element);
        console.log(wrapper);
    });
});