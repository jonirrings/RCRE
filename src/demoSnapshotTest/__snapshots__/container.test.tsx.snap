// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container SnapShot Test container child props parse 1`] = `
<div
  className="rcre-render"
>
  <div
    className="page-container"
  >
    
    <div
      className="page-body"
    >
      <div
        className="rcre-container"
      >
        <div
          className="rcre-container"
        >
          <span
            style={
              Object {
                "lineHeight": "25px",
                "padding": "0 10px",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
